File: leetcode/LC/_c++code/array/0015_three_sum.cpp
#include <climits>
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    
    void twosum(vector<int>& nums, int i, vector<vector<int>> &triplets){
        int start = i +1;
            int end = nums.size() - 1;
            while(start < end){
                
                if(nums[i] + nums[start] + nums[end] == 0){
                    triplets.push_back({nums[i], nums[start], nums[end]});
                    start++;
                    end--;
                    while(start < end && nums[start] == nums[start - 1])
                        start++;
                    while(start < end && nums[end] == nums[end+1])
                        end--;
                }else if(nums[i] + nums[start] + nums[end] > 0){
                    end--;
                }else{
                    start++;
                }
            }   
    }
    
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        sort(nums.begin(), nums.end());
        vector<vector<int>> triplets;
        for(int i = 0; i < nums.size(); i++){
            if(nums[i] > 0){
                break;
            }
            if(i != 0 && nums[i] == nums[i-1]){
                continue;
            }
            twosum(nums, i, triplets);
        }
        return triplets;
    }
};

File: leetcode/LC/array/0015_three_sum.py
from typing import List

class Solution:
    def twoSum(self, nums, i, ans):
        start = i+1
        end = len(nums) - 1
        while(start < end):
            currentSum = nums[start] + nums[end] + nums[i]
            if currentSum == 0:
                ans.append([nums[i], nums[start], nums[end]])
                start += 1
                end -= 1
                # shift end to avoid same start and end values
                while(start < end and nums[end] == nums[end+1]):
                    end -= 1
            elif currentSum > 0:
                end -= 1
            else:
                start += 1

        
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # sort used O(n) space, O(nlogn) time
        nums.sort() 
        ans = []
        for i in range(len(nums)):
            # pick first, if there are duplicates. so, that second one can be used
            if i == 0 or nums[i-1] != nums[i]:
                self.twoSum(nums, i, ans)
        return ans
        